[Gerald Teo] iP

jialatteo opened the pull request:
## **DukePro**
> "Your mind is for having ideas, not holding them." - David Allen [source](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/jialatteo/ip/releases/tag/v0.1)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) 

If you are a Java programmer, you can use it too practice Java too. 
Here's the `main` method:
```java
    public static void main(String[] args) {
        new Duke("tasks.txt").run();
    }

```

Commit:
Add Gradle support

jialatteo committed:
Level-1

jialatteo committed:
Level-2

jialatteo committed:
Level-3

jialatteo committed:
Level-4

jialatteo committed:
Level-5

jialatteo committed:
Level-5

jialatteo committed:
Level-6

jialatteo committed:
Add storage class

jialatteo committed:
Create task file when it does not exist

jialatteo committed:
Merge branch 'branch-Level-7'

jialatteo committed:
Accept dates in yyy-mm-dd format

jialatteo committed:
Merge branch 'branch-Level-8'

jialatteo committed:
Refactor to Ui, Storage, Parser, TaskList classes

jialatteo committed:
Organize classes into packages

jialatteo committed:
Add tasks.txt to .gitignore

jialatteo committed:
Merge branch 'add-gradle-support'

jialatteo committed:
Add settings.gradle

jialatteo committed:
Fix bug where program cannot quit

jialatteo committed:
Set up gradle to run checkstyle test

jialatteo committed:
Add JUnit Tests for Parse and Todo Class

jialatteo committed:
Added JavaDoc comments to classes and methods

jialatteo committed:
Make .gitignore work properly

jialatteo committed:
Merge branch branch-A-JavaDoc

jialatteo committed:
Merge branch branch-A-JavaDoc

jialatteo committed:
Tweak code to comple with coding standard

jialatteo committed:
Merge branch 'A-CodingStandard'

jialatteo committed:
Add function for users to find task with a keyword

jialatteo committed:
Merge branch 'branch-Level-9'

clarence-chew commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,116 @@
+            Scanner sc = new Scanner(file);
+            while (sc.hasNextLine()) {
+                String line = sc.nextLine();
+                String[] splitStr = line.split("\\|");
```
The way this is implemented seems to inconvenience users wanting to use the `|` symbol in their task description.

clarence-chew commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+     * @throws DukeException This tells the user if no such task contains the user's search keyword
+     */
+    private static List<Task> getAllTaskWithKeyword(String str, TaskList taskList) throws DukeException {
+        String[] splitStr = str.split(" ");
```
There's a bit of code duplication here where the idea of "splitting and matching the first word, then getting the rest of the string" is repeated multiple times. Can this common idea be perhaps abstracted away, perhaps with other functions like this?
```java
public static boolean isOfTheCommand(String input, String commandName) {
    String[] splitStr = input.split(" ");
    return splitStr[0].equals(commandName));
}
public static boolean getCommandBody(String input) {
    String[] splitStr = input.split(" ");
    String type = splitStr[0];
    return input.substring(type.length() + 1);
}
```

clarence-chew commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+ * @since 2022-08-28
+ */
+public class Parser {
+    public Parser() {}
```
This seems to be dead code. One way to perhaps make it somewhat useful would be to write it as
`private Parser() {}`, to make it such that the class cannot be instantiated.

clarence-chew commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,43 @@
+     * @return String This returns the string of the event in the specified format
+     */
+    public String toString() {
+        return "[E][" + this.getStatusIcon() + "] " + this.description + " (at: " + at + ")";
```
Do avoid unnecessary use of `this`, as per the coding standard given.

clarence-chew commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    public static void parseCommand(String command, TaskList taskList, Ui ui) {
+        try {
+            if (command.equals("bye")) {
+                ui.quit();
+            } else if (command.equals("list")) {
+                ui.listOutTasks(taskList);
+            } else {
+                String[] splitStr = command.split(" ");
+                Task taskToMark = getTaskToMark(splitStr, taskList);
+                Task taskToAdd = getTaskToAdd(command);
+                Task taskToDelete = getTaskToDelete(splitStr, taskList);
+                List<Task> taskListWithKeyWord = getAllTaskWithKeyword(command, taskList);
+                if (taskToMark != null) {
+                    String action = splitStr[0];
+                    if (action.equals("mark")) {
+                        ui.markTask(taskToMark);
+                    } else {
+                        ui.unmarkTask(taskToMark);
+                    }
+                } else if (taskToAdd != null) {
+                    ui.addTask(taskList, taskToAdd);
+                } else if (taskToDelete != null) {
+                    ui.deleteTask(taskList, taskToDelete);
+                } else if (taskListWithKeyWord.size() != 0) {
+                    ui.printTasksWithKeyword(taskListWithKeyWord);
+                } else {
+                    throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(\n");
+                }
+            }
```
The indentation in this area may be reduced through the use of guard clauses, i.e.
```java
if (command.equals(bye) {
    ui.quit();
    return;
}
if (command.equals(list) {
    ...
    return;
}
String[] splitStr = ...
...
if (taskToMark != null) {
    ...
    return;
}
if (taskToAdd != null) {
    ...
    return;
}
...
throw new DukeException("...");
```

clarence-chew reviewed (commented):
Nice work with your chatbot!

midnightfeverrr commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+     * Constructor for Deadline class
+     * @param description Description of deadline
+     * @param date Due date of deadline
+     * @param isDone Completion status of deadline
```
Comments should have empty line between description and parameter section

midnightfeverrr commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+     * Returns the task's type, completion status, description and due date
+     * @return String This returns the string of the deadline in the specified format
+     */
+    public String toString() {
```
Missing @override tag

midnightfeverrr commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+     * This method takes in the string input of the user
+     * and return the task from the taskList to be marked/unmarked.
+     * When an invalid number is entered, user will get to
+     * enter again
```
Do consider rephrasing comments according to match java coding standard (i.e. MARKS the task from taskList to be done/undone.)

midnightfeverrr commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,116 @@
+ */
+public class Storage {
+
+    private String filePath;
```
consider making the filePath final

cfyjoanna commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+            }
+            return null;
+        } catch (NumberFormatException e) {
+            return null;
```
We should avoid empty catch statements. Maybe you can add a comment explaining why it's okay to return null here?

cfyjoanna commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+        } catch (NumberFormatException e) {
+            return null;
+        }
+    }
```
The code here looks very similar to the one in `getTaskToDelete`. To improve readability, perhaps you could write a separate method to process the checking of valid index?

cfyjoanna commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+        } else if (type.equals("event")) {
+            if (splitStr.length < 2) {
+                throw new DukeException("☹ OOPS!!! The description of an event cannot be empty.\n");
+            }
```
Maybe you could perform the checking of splitStr.length < 2 before checking for the type of task to avoid repeated checking and reduce the length of code. The exception message could be changed to something more generic.

cfyjoanna reviewed (commented):
There are only some minor changes that can be implemented to improve readability. Looks like your code is great otherwise. Nice job! :)

jialatteo committed:
Set up JavaFX

jialatteo committed:
Add images to chat

jialatteo committed:
Add .fxml files

jialatteo committed:
Make chatbot parse commands

jialatteo committed:
Merge branch 'branch-Level-10'

jialatteo committed:
Trim whitespace for commands

jialatteo committed:
Show thrown exception message in chatbot

jialatteo committed:
Use PR reviews from peers to fix code quality and coding standard issues

- Split commands into command type and command body

- Changed to switch case for parseCommand mathod in Parser class to make
  method more readable

- Added a isValidIndex for Parser class for less code duplication

- Made file path of tasks.txt final

jialatteo committed:
Change build.gradle file mainClassName to run jar file

jialatteo committed:
Add Assertion statements

Added assertion statements in Storage, MainWindow and Duke class

jialatteo committed:
Merge pull request #2 from jialatteo/branch-A-Assertions

Add Assertion statements

jialatteo committed:
Improve code quality

- Remove magic numbers
- Remove unused imports/methods

jialatteo committed:
Make code comply with coding standard

jialatteo committed:
Merge pull request #3 from jialatteo/branch-A-CodeQuality

Improve code quality

jialatteo committed:
Add friendlier syntax

Added syntax below:
- "find" = "f"
- "delete" = "rm"
- "todo" = "t"
- "deadline" = "d"
- "event" = "e"
- "bye" = "exit" = "quit"

jialatteo committed:
Make chatbot GUI more aesthetic

jialatteo committed:
Change path of tasks.txt file

jialatteo committed:
Edit javadocs

jialatteo committed:
Add more JavaDoc comments

jialatteo committed:
Add Ui.png to docs folder

jialatteo committed:
Update User guide

jialatteo committed:
Update User guide

jialatteo committed:
Update User guide

jialatteo committed:
Update User guide

jialatteo committed:
Update User guide

jialatteo committed:
Update User guide

jialatteo committed:
Update User guide

jialatteo committed:
Improve coding style

jialatteo committed:
Update User Guide
